PR for milestone-3: Feature/game logic

smcommits opened the pull request:
This pull request is for milstone 3 of the OOP project. In this milestone, the objective was to implement the game logic and make the game functional.

In this milstone, we have implemented the following classes
- Game class for basic game behaviour
- Board class to imititate a board
- Player class to imitiate a player

We have added several helper methods in the classes to implenet the complete game logic.

Commit:
Project Setup

smcommits committed:
Create feature branch, add lib and bin folder, add main.rb

Created feature branch named project-setup
added lib folder using .gitkeep
added main.rb to the bin folder

Commit:
Add /bin/.gitkeep, Make main.rb executable

Commit:
Add /lib/.gitkeep, Make main.rb executable

Added .gitkeep file to lib directory
Made main.rb file executable by changing permission

Commit:
Add /lib/.gitkeep, Make main.rb executable

Added .gitkeep file to lib directory
Made main.rb file executable by changing permission

Commit:
Merge branch 'feature/project-setup' of https://github.com/Lameck1/Tic-Tac-Toe into feature/project-setup

Commit:
Merge pull request #1 from Lameck1/feature/project-setup

Milestone 1: Project-Setup

smcommits committed:
Initialize user inteface, Add User interface componenets

Intitialized the feature branch user interface
Added the componenets relating to user interface in the main.rb file
- Added a method to diplay instructions
- Added a mehtod to take input
- Added a method to display board

Commit:
Finalize dummy logic

Added class Board
- Added helper methods to board class
- Helper methods return nil
Added Player class
- Added switch_player method
- switch_player return nil
Initialize dummy logic in main.rb using dummy classes

smcommits committed:
Update dummy logic

Update dummy logic to depict the entire logic of the game
Returned to user_input to position
Added update_board method
Added ininite loop with a break clause of win? and tied? method

Commit:
Fix Linter Errors

Refactor explicit return statement to implicit
Refactor indentation

smcommits committed:
Add win method, update method, get_name method, get_user_input method

Added a win method to decide if someone has won
Added an update method to update the board
Added the get_name method to get the name of the users
Added the get_user_input method to get the psoitonal inputs from the user

Commit:
Fix Linter Errors, Refactor get_name method

Fixed indentation and removed trailing whitespaces
Changed method name to user_name

smcommits committed:
Add dummy methods, update interface in main.rb

Added dummy methods
 - position_taken?
 - valid_move?
 - update_board

 Added interface to main.rb
 - Added input validation
 - Moved draw method from board library to main.rb

smcommits committed:
Merge pull request #2 from Lameck1/feature/user-interface

PR for second milstone: user interface

Commit:
Add draw, valid_move? and position_taken? methods

draw method, draws the board
valid_move validates players move on the board
position_taken prevents re-usage of board cells

smcommits committed:
Add method: win, update, tied, and switch_player to board class

The win method checks if any winning combination is true
The tied method checks if all the cells are occupied
The Update method updates the baord after user input
The switch_player method switched turns at each loop

Commit:
Update player class and initialized game class

Added followign methods to the game class
- display_instructions to display game intructions
- user_name for getting player names
- get_user_input to get players moves on the board

smcommits committed:
Add methods: validate, display_details, and play to game class

- Validate methods validates the user input for a move
- Display details display player details
- Play intializes one instance of a game

Commit:
Refactor main.rb

Deleted methods moved to classes

smcommits committed:
Add name_valid method, update user_name class, added ascii

Added a valid_name method to game class, to verify if the name entered by user is not blank
Updated user_name method to have name_valid method
Added greeter ASCII

mimipeshy commented:
## STATUS : CHANGES REQUIRED :red_circle: 

Hi team :wave: , this game looks AMAZING :smile: . I like the fact that you incorporated the display, your game is easy to follow along, it works well. However, before approving lets make some changes :computer: 

- [x]  No `puts/print` and `get.chomp `outside` main.rb` file.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- [x]  The method `def valid name?` on game.rb if i enter two blank spaces, i am able to create a blank user. Maybe check on how to eliminate the extra spaces
- [x] DRY rule should be respected.([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/dry.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
in this case on game.rb, just create a method whose work is to display the tic-tac-toe and re-use it as much as you wish. The goal is to have one method do just one thing.

That's all :smile:  KIndly submit another review when done :smile: 
Regards 
Peris Ndanu
Feel free to reach out on slack if you have any queries

smcommits committed:
Implement change request

Refactored the display of Tic-Tac-Toe in display_banner method
Removed gets.chomp and puts from every file and placed them inside main.rb
Redesigned the valid_name method to fix the bug of accepting empty spaces

descholar-ceo reviewed (approved):
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:  
## :clap: :clap: :clap: Congratulations @Lameck1 and @smcommits  , your project is approved :tada: :tada: :tada: 
You have implemented all requested changes for this project to mark it as complete, before your merge the changes to the base branch, remember to remove the `.gitkeep` file from lib folder, you no longer need it now. And also you can read [this documentation](https://rdoc.info/github/fazibear/colorize) to know how to `install` and `use` [colorize](https://rubygems.org/gems/colorize) as a gem, no need to add the whole colorize project / folder into your project. 

Don't forget to fill the mandatory project completion form, and see you to next projects,

Happy coding :computer: :computer: :computer: 

Re-reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)

smcommits commented:
@descholar-ceo Thank you so much for the feedback! 
I will go ahead and remove the .gitkeep file. Also, is there a way we can use an installed ruby gem like _colorize_ in a project so that users don't need to install it on other machines? Please, let me know if you have some information about that. 
Thanks again!
